### ISSUE TITLE
Security Vulnerability: Privilege Escalation via Content Provider (CVE-2018-9492)

### ISSUE THREAD

**Issue by vestrel00** [getsentry/sentry-java#2460]
### Integration

sentry-android

### Build System

Gradle

### AGP Version

7.3.0

### Proguard

Enabled

### Version

6.6.0

### Steps to Reproduce

1. Integrate Android sentry into an application
2. Scan the app using [Secure theorem](https://www.securetheorem.com/) for security vulnerabilities.

### Expected Result

A security issue should not exist.

### Actual Result

[Secure theorem](https://www.securetheorem.com/) is flagging a security issue "Privilege Escalation via Content Provider" ([CVE-2018-9492](https://www.cvedetails.com/cve/CVE-2018-9492/)).

### More info

We are currently using sentry version [6.6.0](https://github.com/getsentry/sentry-java/releases/tag/6.6.0), which was release on Oct 24, 2022. 

We routinely upgrade our sentry version, so we will be upgrading to the latest; ([6.11.0](https://github.com/getsentry/sentry-java/releases/tag/6.11.0)). However, I doubt that the security issue is fixed in more recent versions, given that the security issue has existed for over 2 years now on our app and I don't see any issues related to this being fixed in the [sentry issues page](https://github.com/getsentry/sentry-java/issues?q=is%3Aissue+). We are just now being forced to resolve this 3rd party security issue due to our company's efforts in increasing app security. Hence, me creating this issue 😅 

### About the security vulnerability

Here are all the _copy-paste_ information we have about the security issue from secure theorem.

Note that we cannot fix the issue by setting our app's `minSdkVersion` to at least 29. Therefore, we need this security issue to be fixed within sentry itself 🙏 

#### AFFECTED CODE

<img width="828" alt="Screenshot 2023-01-09 at 9 55 33 AM" src="https://user-images.githubusercontent.com/4332753/211396670-398b1dd4-7790-4825-a100-517506028d2d.png">

#### DESCRIPTION

Unpatched versions of Android 8 and 9 and vulnerable to [CVE-2018-9492](https://www.cvedetails.com/cve/CVE-2018-9492/): Privilege Escalation Vulnerability in [ActivityManagerService.java](https://android.googlesource.com/platform/frameworks/base/+/962fb40991f15be4f688d960aa00073683ebdd20). This vulnerability allows any third party malicious App to access private data of a victim App without any user interaction.

Any third party App can acquire read/write access to any ContentProvider in the victim App. It doesn't matter if the ContentProvider is exported/non-exported or if the ContentProvider is protected with permissions.

[Security patches](https://source.android.com/security/bulletin/2018-10-01) were released by Google in 2018 to mitigate this vulnerability. Majority of the devices should be patched by now. However, it is possible that some device manufacturers may not have pushed these updates to the users or that some users may have chosen to not install the security updates, leaving a very small percentage of users vulnerable to this issue.

#### RECOMMENDATION

This issue is a vulnerability in the Android OS and can't be exploited on a patched version of Android. However, if you are concerned about unpatched devices, the following options can be used to mitigate the vulnerability:

Option 1: Validate the App requesting for access to the ContentProvider.

Get the calling App's package name via the [getCallingPackage()](https://developer.android.com/reference/android/content/ContentProvider#getCallingPackage()) API. You can use this information to validate that the recipient is allowed to receive the data.

Option 2: Set minSdkVersion to at least 29 for your application.

This vulnerability affects unpatched versions of Android 8 and 9 (API level 26 to 28), by setting the minSdkVersion to 29 your App will install only on devices running Android 10 and above. However, please note that doing so will cause your App not to work on a majority of Android devices.

#### SECURE CODE

```kotlin
class TestProvider extends android.content.ContentProvider {
    @Override
    public boolean onCreate() {
        return false;
    }

    private void securityCheck(Uri uri)  {
        String callingPackage = this.getCallingPackage();
        if(callingPackage != null && (callingPackage.equals("com.trusted.app"))) {
            return;
        }
       SecurityException exception = new SecurityException("Provider does not allow granting of Uri permissions");
       throw exception;
    }

    @Nullable
    @Override
    public Cursor query(@NonNull Uri uri, @Nullable String[] projection, @Nullable String selection, @Nullable String[] selectionArgs, @Nullable String sortOrder) {
        securityCheck(uri);
        return null;
    }

}
```

**Comment by romtsn** [comment]
Hm, but if we are not doing anything with the Uri, why does it get flagged? The sole purpose of the ContentProviders in our code is to automatically [init](https://github.com/getsentry/sentry-java/blob/main/sentry-android-core/src/main/java/io/sentry/android/core/SentryInitProvider.java) the SDK inside `onCreate`, but that's it.

Also, would [removing the content providers](https://docs.sentry.io/platforms/android/configuration/manual-init/) and using manual init help?

I'm a bit hesitant adding the suggested workaround, because this is the first time this issue is reported, and looks like a false positive actually.

**Comment by romtsn** [comment]
@vestrel00 thanks for the report nonetheless!

Please, beware that 6.6.0 has a bug (through 6.8.0 where it has been fixed), where the cached events/transactions might fail to get sent when using manual init, so I'd strongly advise you to update the SDK to the later version (6.8.0+). Sorry for the trouble!

**Comment by vestrel00** [comment]
> Please, beware that 6.6.0 has a bug (through 6.8.0 where it has been fixed), where the cached events/transactions might fail to get sent when using manual init, so I'd strongly advise you to update the SDK to the later version (6.8.0+). Sorry for the trouble!

Wow! Thanks so much for letting me know! I will upgrade to the latest Sentry version as part of my work =)

**Comment by artour-bakiev** [comment]
Hey @romtsn. I'm from the same organisation as @vestrel00 and just want to add my 5 cents regarding this issue.

You've mentioned that
> we are not doing anything with the Uri

Unfortunately it's not a marker to be safe from [this issue](https://www.cvedetails.com/cve/CVE-2018-9492/) perspective. It looks like the attack actually may happen **after** after a public method of the provider is called.
The information provided by https://www.cvedetails.com/cve/CVE-2018-9492/ doesn't look sufficient to reveal the attack details.

So it seems we should assume it's a security breach unless we prove the opposite.

**Comment by romtsn** [comment]
@artour-bakiev @vestrel00 would you be up for opening a PR attempting to fix this issue? I'm unsure we can reliably fix the issue without a proper reproducer. Since you already have the security tool running, perhaps you could take a stab at checking if the proposed workaround suppresses the vulnerability?

**Comment by romtsn** [comment]
Also, even if this is not a false positive and an attacker can get access to our content provider, there's no data to access at all, unless I'm misunderstanding something

**Comment by vestrel00** [comment]
Thanks for digging deeper and being diligent @artour-bakiev! I took a closer look at https://www.cvedetails.com/cve/CVE-2018-9492/ and found that there could indeed be a vulnerability and this may not be a false positive! The following is very alarming...

<img width="967" alt="Screen Shot 2023-01-11 at 10 56 00 AM" src="https://user-images.githubusercontent.com/4332753/211919942-9615296d-96ec-4158-a620-5cf310310ec6.png">

> would you be up for opening a PR attempting to fix this issue? I'm unsure we can reliably fix the issue without a proper reproducer. Since you already have the security tool running, perhaps you could take a stab at checking if the proposed workaround suppresses the vulnerability?

My pleasure! I am honored to! This is a chance to contribute to an awesome OSS, so I'm super excited to get this opportunity 🔥 ❤️ 

So, I created a PR with the probable fix with complete information; https://github.com/getsentry/sentry-java/pull/2466

Please take a look @romtsn !

**Comment by markushi** [comment]
Quick update: The fix will land in the official `6.12.0` release, which should be available within the next hours.
@vestrel00 @artour-bakiev thanks for your contributing, much appreciated! 🚀
